idx/memory/tag_query.go

```diff
@@ -0,0 +1,411 @@
+			}
+			break
+		}
+
+		// disallow ; in key
+		if expr[pos] == 59 {
+			return expression{operator: PARSING_ERROR}
+		}
+	}
+
+	key := expr[:pos]
+
+	// if !
+	if expr[pos] == 33 {
+		not = true
+		pos++
+	}
+
+	// expecting a =
+	if expr[pos] != 61 {
+		return expression{operator: PARSING_ERROR}
+	}
+	pos++
+
+	// if ~
+	if len(expr) > pos && expr[pos] == 126 {
+		regex = true
+		pos++
+	}
+
+	valuePos := pos
+	for ; pos < len(expr); pos++ {
+		// disallow ; in value
+		if expr[pos] == 59 {
+			return expression{operator: PARSING_ERROR}
+		}
+	}
+	value := expr[valuePos:]
+
+	if not {
+		if regex {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: NOT_MATCH}
+		} else {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: NOT_EQUAL}
+		}
+	} else {
+		if regex {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: MATCH}
+		} else {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: EQUAL}
+		}
+	}
+}
+
+func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+	q := TagQuery{from: from}
+
+	if len(expressions) == 0 {
+		return q, errInvalidQuery
+	}
+
+	for _, expr := range expressions {
+		e := parseExpression(expr)
+		if e.operator == PARSING_ERROR {
+			return q, errInvalidQuery
+		}
+
+		// special case of empty value
+		if len(e.value) == 0 {
+			if e.operator == EQUAL || e.operator == MATCH {
+				q.notMatch = append(q.notMatch, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			} else {
+				q.match = append(q.match, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			}
+		} else {
+			// always anchor all regular expressions at the beginning
+			if (e.operator == MATCH || e.operator == NOT_MATCH) && len(e.value) > 0 && e.value[0] != byte('^') {
+				e.value = "^" + e.value
+			}
+
+			switch e.operator {
+			case EQUAL:
+				q.equal = append(q.equal, e.kv)
+			case NOT_EQUAL:
+				q.notEqual = append(q.notEqual, e.kv)
+			case MATCH:
+				q.match = append(q.match, e.kv)
+			case NOT_MATCH:
+				q.notMatch = append(q.notMatch, e.kv)
+			}
+		}
+	}
+
+	return q, nil
+}
+
+func (q *TagQuery) getInitialByMatch(index TagIndex) (int, TagIDs, error) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key that has the smallest number of values because we want to
+	// reduce the number of matches and start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.match {
+		l := len(index[q.match[i].key])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// shortcut if pattern is ^.+ (f.e. expression "key!=~" will be translated to "key=~^.+")
+	if q.match[startId].value == "^.+" {
+		for _, ids := range index[q.match[startId].key] {
+			for id := range ids {
+				resultSet[id] = struct{}{}
```

Comment:
why don't we verify that these id's actually have a tag that would match the pattern (e.g. have a len > 0). is it simply because any tag value *must* be a non-empty string? 
if this is the case, please add that as a comment.

Reply:
yes, exactly that. every tag must have a value, and this pattern will match any value. going to add a comment